strings/check_anagram.nim

```diff
@@ -0,0 +1,179 @@
+## Check Anagram
+## =============
+## wiki: https://en.wikipedia.org/wiki/Anagram
+##
+## Two words are anagrams if:
+## - They are made up of the same letters.
+## - The letters are arranged differently.
+## - The case of the characters in a word is ignored.
+## - They are not the same word (a word is not an anagram of itself).
+##
+## A word is any string of characters `{'A'..'Z', 'a'..'z'}`. Other characters,
+## including whitespace, numbers and punctuation, are considered invalid and
+## raise a `ValueError` exception.
+##
+## Note: Generate full doc with `nim doc --docinternal check_anagram.nim`
+
+runnableExamples:
+  doAssert "coder".isAnagram("credo")
+
+  doAssert not "Nim".isAnagram("Nim")
+
+  doAssert not "parrot".isAnagram("rapport")
+
+  let raisedException = try:
+      discard "Pearl Jam".isAnagram("Maple Jar")
+      false
+    except ValueError:
+      true
+  doassert raisedException
+
+## Tests
+## -----
+## This module includes test suites for each public function, which will run if
+## the module is compiled as an executable.
+##
+## TODO
+## ----
+##
+## - Unicode version of the algorithm.
+## - Check if the arguably more idiomatic solution using whole-word operations
+##   is beneficial or detrimental to the performance, compared to char-by-char
+##   approach of this module.
+
+type
+  Map = array[range['a'..'z'], Natural] ## An associative array with a direct
+                                        ## mapping from Char to a counter slot
+
+const
+  UpperAlpha = {'A'..'Z'}
+  LowerAlpha = {'a'..'z'}
+
+func toLowerUnchecked(c: char): char {.inline.} =
+  ## Lowers the case of an ASCII uppercase letter ('A'..'Z').
+  ## No checks performed, the caller must ensure the input is a valid character.
+  ##
+  ## See also:
+  ## * `strutils.toLowerAscii <https://nim-lang.org/docs/strutils.html#toLowerAscii%2Cchar>`_
+  assert c in UpperAlpha
+  # The difference between numerical values for chars of uppercase and
+  # lowercase alphabets in the ASCII table is 32. Uppercase letters start from
+  # 0b100_0001, so setting the sixth bit to 1 gives letter's lowercase pair.
+  char(uint8(c) or 0b0010_0000'u8)
+
+template normalizeChar(c: char) =
+  ## Checks if the character is a letter and lowers its case.
+  ##
+  ## Raises a `ValueError` on other characters.
+  if c in LowerAlpha: c
+  elif c in UpperAlpha: toLowerUnchecked(c)
+  else: raise newException(ValueError, "Character '" & c & "' is not a letter!")
+
+func isAnagram*(a, b: openArray[char]): bool {.raises: ValueError.} =
+  ## Checks if two words are anagrams of one another.
+  ##
+  ## Raises a `ValueError` on any non-letter character.
+  if a.len != b.len: return false
+  var seenDifferent = false
+  var mapA, mapB: Map
+  for chIdx in 0..<a.len:
```

Comment:
a, b single variable names :(
What are they ? Words. Why not wordA, wordB instead?
Apart from this style-cliving discussion, it seems to me that if they are openArrays, they may be indexed by something else: prefer a.low and a.high.
What if a and b are indexed differently ? Yes, you may have to do two additional sums when accessing b: b[low(b)-low(a)+chIdx]. We may suppose here that low(a) == low(b).

Reply:
Here I'm following the Standard Library convention of naming arguments.

I'm not really sure what are you up to regarding indexing:

```nim
let a: array[range[10..13], char] = ['a', 'b', 'c', 'd']
let b: array[range[0..3], char] = ['w', 'x', 'y', 'z']

proc foo(a, b: openArray[char]) =
  for i in 0..3: echo a[i], b[i]

foo(a, b)
```

Reply:
I thought that array indexes were passed along with the openArray into functions, sorry about this then.
I see nothing in NEP1 about single variable names, but I am nitpicking here.

Reply:
I kind of understand the logic why the standard library uses `a`, `b`... for naming equivalent entities (where there's no clear relationship between them, unlike `src`-`dst` or `s`-`sub`), like in various `cmp` procs:
* It conveys these are the main subjects of the proc.
* It shows we don't know anything about the values beside its type.
* It conveys it's the raw input, not parsed/validated and frees the basic names like `word` to be used after that.
* It's generic so it works for this pattern regardless of the type.

When I start thinking about it, from the logic above follows in this case it's better to be explicit we expect not just any string. So I've talked myself into changing the names. :P

Reply:
Well, my main reasoning is the func may fail on invalid input, so arg naming should reflect the kind of data expected: "Do not put any old string, it has to be a word!".

> Here I understand that the input has been stripped down of its extra characters.

I mean, we can shadow args in Nim of course, but even if we couldn't: `let month = a.parseMonth().toLower()` `a` is just some data which can be a `month`, means the proc has to deal with it in some way. Of course, it's all moot when you remember you should use types for this, but the complexity would quickly grow in case of a bunch of interleaved procs working on raw data (one example being `strutils`). So the correct approach always depends on the context of code usage.